action_server/src/sema4ai/action_server/_api_action_routes.py

```
@@ -0,0 +1,340 @@
+
+
+class _ActionRoutes:
+    def __init__(
+        self,
+        whitelist: str | None,
+        endpoint_dependencies: list[params.Depends],
+    ):
+        """
+        Args:
+            whitelist: The whitelist of actions to register.
+            endpoint_dependencies: The dependencies for the endpoint (which will require the API key).
+            api_key: The API key to use for the endpoint.
+        """
+        from mcp.server.sse import SseServerTransport
+        from mcp.server.streamable_http_manager import StreamableHTTPSessionManager
+
+        from sema4ai.action_server._models import Action, ActionPackage
+        from sema4ai.action_server.mcp.setup_mcp_server_from_actions import (
+            McpServerSetupHelper,
+        )
+
+        self.whitelist = whitelist
+        self.endpoint_dependencies = endpoint_dependencies
+        self.action_package_id_to_action_package: dict[str, ActionPackage] = {}
+        self.actions: list[Action] = []
+        self.registered_route_names: set[str] = set()
+        self.mcp_server_setup_helper: McpServerSetupHelper = McpServerSetupHelper()
+        self._session_manager: StreamableHTTPSessionManager | None = None
+        self._sse_transport: SseServerTransport | None = None
+
+    def setup_mcp_server(self, api_key: str | None) -> None:
+        """
+        Setup the MCP server with both streamable HTTP and SSE support.
+        """
+        from starlette.middleware.authentication import AuthenticationMiddleware
+
+        from sema4ai.action_server._app import get_app
+
+        app = get_app()
+
+        middleware: list[Middleware] = []
+        if api_key:
+            middleware.append(
+                Middleware(
+                    AuthenticationMiddleware,
+                    backend=APIKeyAuthBackend(api_key=api_key),
+                )
+            )
+
+        self._setup_mcp(app, middleware)
+        self._setup_sse(app, middleware)
```

It looks like we're still doing the setup for MCP + SSE if the API key is None.
If the authentication is mandatory for those, maybe throw an exception?

---

It's not mandatory, if the API key is not set, we should still setup the mcp server.